include/swift/SIL/SILInstruction.h

```diff
@@ -8087,8 +8087,25 @@ class GraphOperationResult final : public MultipleValueInstructionResult {
   }
 };
 
-/// A graph operation. This instruction will be extracted to a graph program
-/// via graph program extraction passes.
+/// A graph operation, which takes operands and attributes and returns results.
+///
+/// Operands have values that are possibly unknown at compile time. Operands
+/// are grouped into `GraphOperationInfo::StructuredArgument`s. See there for
+/// more documentation. This structure is mangled into `Name`.
+///
+/// Attributes have values that are known at compile time, and these values are
+/// stored in the GraphOperationInst.
+///
+/// Results can be represented in 3 different ways:
+/// 1. As an output parameter with type that is a TensorFlow value or aggregate
+///    of TensorFlow values.
+/// 2. As a single result that is a TensorFlow value or aggregate of TensorFlow
+///    values.
+/// 3. As multiple results, where each result is a TensorFlow value (not an
+///    aggregate of TensorFlow values!).
+/// The later result representations are "better" than the earlier ones because
+/// code performing the operations has to do less work to deal with the results.
+/// Various compiler passes try to improve the result representations.
```

Comment:
nice!

it would also be useful to document the guaranteed output from deabstraction pass. e.g. will we still have case `#2` for results, or will we only have #3? If the latter, graph lowering would have an easier time (same for the new IRgen code path).

Reply:
Right now, deabstraction guarantees 3.

But if we want to IRGen functions where the results are generic, then IRGen will have to deal with 1! This is related to https://bugs.swift.org/browse/SR-8830. It probably makes sense to handle the easy case (3) first, and then add support for generic results in IRGen later.

I'll document the current behavior here.